test/unit/instance-id/instance-id.spec.ts

```diff
@@ -0,0 +1,193 @@
+    before(() => utils.mockFetchAccessTokenRequests());
+  
+    after(() => nock.cleanAll());
+  
+    beforeEach(() => {
+      mockApp = mocks.app();
+      mockCredentialApp = mocks.mockCredentialApp();
+      iid = new InstanceId(mockApp);
+
+      gcloudProject = process.env.GCLOUD_PROJECT;
+
+      nullAccessTokenClient = new InstanceId(mocks.appReturningNullAccessToken());
+      malformedAccessTokenClient = new InstanceId(mocks.appReturningMalformedAccessToken());
+      rejectedPromiseAccessTokenClient = new InstanceId(mocks.appRejectedWhileFetchingAccessToken());
+    });
+  
+    afterEach(() => {
+      process.env.GCLOUD_PROJECT = gcloudProject;
+      return mockApp.delete();
+    });
+  
+  
+    describe('Constructor', () => {
+      const invalidApps = [null, NaN, 0, 1, true, false, '', 'a', [], [1, 'a'], {}, { a: 1 }, _.noop];
+      invalidApps.forEach((invalidApp) => {
+        it('should throw given invalid app: ' + JSON.stringify(invalidApp), () => {
+          expect(() => {
+            const iidAny: any = InstanceId;
+            return new iidAny(invalidApp);
+          }).to.throw('First argument passed to admin.instanceId() must be a valid Firebase app instance.');
+        });
+      });
+  
+      it('should throw given no app', () => {
+        expect(() => {
+          const iidAny: any = InstanceId;
+          return new iidAny();
+        }).to.throw('First argument passed to admin.instanceId() must be a valid Firebase app instance.');
+      });
+
+      it('should throw given an invalid credential without project ID', () => {
+        // Project ID not set in the environment.
+        delete process.env.GCLOUD_PROJECT;
+        expect(() => {
+          return new InstanceId(mockCredentialApp);
+        }).to.throw(noProjectIdError);
+      });
+  
+      it('should not throw given a valid app', () => {
+        expect(() => {
+          return new InstanceId(mockApp);
+        }).not.to.throw();
+      });
+    });
+  
+    describe('app', () => {
+      it('returns the app from the constructor', () => {
+        // We expect referential equality here
+        expect(iid.app).to.equal(mockApp);
+      });
+  
+      it('is read-only', () => {
+        expect(() => {
+          (iid as any).app = mockApp;
+        }).to.throw('Cannot set property app of #<InstanceId> which has only a getter');
+      });
+    });
+
+    describe('deleteInstanceId()', () => {
+
+      // Stubs used to simulate underlying api calls.
+      let stubs: sinon.SinonStub[] = [];
+      const expectedError = new FirebaseInstanceIdError(InstanceIdClientErrorCode.API_ERROR);
+      const testInstanceId = 'test-iid';
+
+      afterEach(() => {
+        _.forEach(stubs, (stub) => stub.restore());
+        stubs = [];
+      });
+
+      it('should be rejected given no instance ID', () => {
+        return (iid as any).deleteInstanceId()
+        .should.eventually.be.rejected.and.have.property('code', 'instance-id/invalid-instance-id');
```

Comment:
Indent second line `.should...` and `.then` below too.

Reply:
Done